### ISSUE TITLE
Handling NaNs in predictions

### ISSUE THREAD

**Issue by abdulfatir** [awslabs/gluon-ts#2560]
## Description
Currently, the default behavior in the evaluator is to skip NaN values when aggregating metrics. This results in the unexpected behavior where a model that produces NaNs as forecasts for a subset of time series may seem to perform better (particularly when looking at the `ND` and `mean_wQuantileLoss`) metrics. In the following snippet, I show how injecting NaNs into forecast "improves" the model performance. Here's what I do in the snippet:

- Create a test dataset with two time series (sine and cosine wave).
- Evaluate using `MeanPredictor` (the predictor doesn't really matter here).
- Inject NaNs in all forecast samples for the first time series.
- The `ND` and `mean_wQuantileLoss` improves (almost by half).
- Inject NaNs in all forecast samples for the second time series too.
- The `ND` and `mean_wQuantileLoss` is now `0.0` and other metrics are NaN. This is because the aggregation function for MSE-like metrics is `mean` whereas for the components of `ND` and `mean_wQuantileLoss` it is `sum` and:

```
pd.Series([float("nan")]).agg("sum", skipna=True) # equals 0.0
pd.Series([float("nan")]).agg("mean", skipna=True) # equals NaN
```

I'm not exactly sure what should the behavior be in the case when forecast contains NaNs; however, my suggestion is as follows.
- Add a `strict` parameter to the `Evaluator`.
- If `strict` is `True`, raise an error when the forecast contains invalid values.
- If `strict` is `False`, raise a warning when the forecast contains invalid values. 

Thanks to @marcelk1102 for initially spotting this issue. 

cc @lostella @dcmaddix 

## To Reproduce
(Please provide minimal example of code snippet that reproduces the error. For existing examples, please provide link.)

```python
In [1]: import numpy as np
   ...: import pandas as pd
   ...: from gluonts.model.trivial.mean import MeanPredictor
   ...: from gluonts.evaluation import make_evaluation_predictions, Evaluator

In [2]: x = np.arange(7 * 24)
   ...: y = np.vstack([np.sin(x * 2 * np.pi / 24), np.cos(x * 2 * np.pi / 24)])
   ...: y = y + np.random.randn(*y.shape) * 0.1
   ...: test_dataset = [{"target": yi, "start": pd.Period("2023-01-01", freq="H")} for yi in y]

In [3]: predictor = MeanPredictor(prediction_length=24, context_length=24)
   ...: forecast_iter, test_iterator = make_evaluation_predictions(test_dataset, predictor)
   ...: forecast_list = list(forecast_iter)
   ...: test_list = list(test_iterator)

In [4]: evaluator = Evaluator()
   ...: agg_metrics, _ = evaluator(test_list, forecast_list, num_series=len(test_dataset))
   ...: agg_df = pd.DataFrame([agg_metrics])[["MSE", "NRMSE", "ND", "mean_wQuantileLoss"]]
   ...: print(agg_df)
Running evaluation: 100%|████████████████████████████████████████████████████| 2/2 [00:00<00:00, 28.71it/s]
        MSE     NRMSE        ND  mean_wQuantileLoss
0  0.523586  1.108964  1.017709            0.725236

In [5]: forecast_list[0].samples[:, 0] = float("nan")
   ...: agg_metrics, _ = evaluator(test_list, forecast_list, num_series=len(test_dataset))
   ...: agg_df = pd.DataFrame([agg_metrics])[["MSE", "NRMSE", "ND", "mean_wQuantileLoss"]]
   ...: print(agg_df)
Running evaluation: 100%|████████████████████████████████████████████████████| 2/2 [00:00<00:00, 33.07it/s]
        MSE     NRMSE        ND  mean_wQuantileLoss
0  0.531871  1.117704  0.509783            0.370515

In [6]: forecast_list[1].samples[:, 0] = float("nan")
   ...: agg_metrics, _ = evaluator(test_list, forecast_list, num_series=len(test_dataset))
   ...: agg_df = pd.DataFrame([agg_metrics])[["MSE", "NRMSE", "ND", "mean_wQuantileLoss"]]
   ...: print(agg_df)
Running evaluation: 100%|████████████████████████████████████████████████████| 2/2 [00:00<00:00, 44.72it/s]
   MSE  NRMSE   ND  mean_wQuantileLoss
0  NaN    NaN  0.0                 0.0
```

## Error message or code output
(Paste the complete error message, including stack trace, or the undesired output that the above snippet produces.)

```
put error or undesired output here
```

## Environment
- Operating system:
- Python version:
- GluonTS version: 0.11.7
- MXNet version:

(Add as much information about your environment as possible, e.g. dependencies versions.)

**Comment by lostella** [comment]
As agreed offline, it currently doesn't make much sense for a model to output a `nan`: as long as `nan` encodes "missing", and instead all admissible values for the series are real numbers, then the model should output real numbers.

So something different should definitely happen in the evaluation. I think we could consider the following options:
1. Raising an error when `nan` is encountered in forecasts (fail evaluation);
2. "Discarding" evaluation instances where `nan` occurs, not accounting for them in aggregate metrics, raise a warning, and report the number of failed instances as part of the aggregate metrics.

Item 1 could be the default, and have the option to enable 2. @dcmaddix @abdulfatir @jaheba what's your opinion here?

Of course in parallel we should make sure that:
3. The seasonal naive method has some fall-back behaviour when `nan` is encountered, so that it always produces a "valid" prediction.

These should be 3 separate PRs, with 1 and 3 having the priority in my opinion.

**Comment by jaheba** [comment]
I agree on optionally allowing `nan` values in forecasts. In some way predicting a missing value can be seen as more "honest" than making up some other value.

To me it makes sense to always allow missing values in the evaluation -- this should be a property which is checked in the predictor and not code using it.

**Comment by abdulfatir** [comment]
@jaheba Sorry, I missed this conversation before submitting the PR. Why do you think we should not check this in the evaluator? In my opinion, we should have this check in the evaluator for the following reason: If an end user just uses the evaluation from GluonTS and their model outputs NaNs for some reason (which in most cases is not the expected behavior), they will get superior performance metrics which is incorrect.

**Comment by jaheba** [comment]
I guess I'm convinced now :)

If we understand the evaluator as some kind of authority grading forecasts, it should also decide what values to accept.

We would also need to add this test to the new `ev` module.

**Comment by abdulfatir** [comment]
Thanks @jaheba! I just checked the `ev` module and it doesn't seem to work with `Forecast` objects but rather with numpy arrays directly, so I am not sure where exactly to add the validation. Any suggestions?

**Comment by lostella** [comment]
Resolved by #2568
